### DISCUSSION TITLE
Plans for forking `Nav2` and `Moveit2`

### DISCUSSION THREAD

**Discussion by franklinselva** [space-ros/space-ros#189] [Ideas]
The current ROS Distro we are working on is **Humble** and we are planning to support **Jazzy** at the moment. I would like to see if there are any plans for forking a version of `nav2` and `moveit2` specifically for SpaceROS.

**Comment by Bckempa** [comment]
There has been talk about a "Space Moveit" and "Space Nav" in the past, but I would be seriously hesitant to take that on; we don't have enough active contributors to shoulder that technical burden. Can you clarify a bit more about what you would expect from such a fork?

This might be more appropriate as a discussion, we can migrate that over if others agree.

This is relevant to #157

**Comment by franklinselva** [comment]
I wanted to see if we are on the same page. I agree with what you have mentioned. `Nav2` and `Moveit2` are already large projects with its ecosystem and contributors. I think it's easier to offload some work to these communities instead of forking and maintaining it. 

The reason for asking this is to find the reason for building the `nav2` and `moveit2` from the source code. To explain this, I have seen that the [ROS apt key](https://github.com/space-ros/space-ros/blob/57649e50be93faa494777e70308c95bb19a7338e/Earthfile#L52) is already a part of the base docker image. This means we do not have to build nav2 and moveit2 from sources as we do in spaceros/docker ([here](https://github.com/space-ros/docker/tree/main/navigation2) and [here](https://github.com/space-ros/docker/tree/main/moveit2))

This is the dockerfile I came up with for `nav2`.

```Dockerfile
FROM osrf/space-ros:core

# Define arguments used in the metadata definition
ARG VCS_REF
ARG VERSION="preview"
ARG NAV2_BRANCH=${ROS_DISTRO}
ARG NAVIGATION2_WS=${HOME}/nav2_ws

# Specify the docker image metadata
LABEL org.label-schema.schema-version="1.0"
LABEL org.label-schema.name="Space ROS Navigation2"
LABEL org.label-schema.description="Preview version of the Navigation2/Space ROS platform"
LABEL org.label-schema.vendor="Open Robotics"
LABEL org.label-schema.version=${VERSION}
LABEL org.label-schema.url="https://github.com/space-ros"
LABEL org.label-schema.vcs-url="https://github.com/space-ros/stacks"
LABEL org.label-schema.vcs-ref=${VCS_REF}

# Install dependencies
RUN sudo apt update \
  && sudo apt install -y ros-${ROS_DISTRO}-nav2-amcl \
  ros-${ROS_DISTRO}-nav2-behavior-tree \
  ros-${ROS_DISTRO}-nav2-behaviors \
  ros-${ROS_DISTRO}-nav2-bt-navigator \
  ros-${ROS_DISTRO}-nav2-common \
  ros-${ROS_DISTRO}-nav2-core \
  ros-${ROS_DISTRO}-nav2-costmap-2d \
  ros-${ROS_DISTRO}-nav2-dwb-controller \
  ros-${ROS_DISTRO}-nav2-graceful-controller \
  ros-${ROS_DISTRO}-nav2-lifecycle-manager \
  ros-${ROS_DISTRO}-nav2-map-server \
  ros-${ROS_DISTRO}-nav2-msgs \
  ros-${ROS_DISTRO}-nav2-navfn-planner \
  ros-${ROS_DISTRO}-nav2-mppi-controller \
  ros-${ROS_DISTRO}-nav2-planner \
  ros-${ROS_DISTRO}-nav2-smoother \
  ros-${ROS_DISTRO}-nav2-velocity-smoother \
  ros-${ROS_DISTRO}-interactive-markers

# TODO: Install slam_toolbox

RUN echo "source /opt/ros/${ROS_DISTRO}/setup.bash" >> ${HOME}/.bashrc

# Flush cache
RUN sudo apt clean \
  && sudo rm -rf /var/lib/apt/lists/* \
  && sudo rm -rf /tmp/*

CMD ["bash"]
```
This is the same case for `moveit2`.

I am sure there will be use cases where we need a special planner or control on navigation and manipulation for space applications. This can be maintained in the space-ros organization and used in the docker image directly.

Taking this further, I think we don't need to keep the changes related to `nav2` and `moveit2` at docker/stacks repository anymore if this is the case. This level of change is easier to maintain from the main repository and less prone to issues related to different ROS Distros.

**Comment by EzraBrooks** [comment]
The source builds are for two main reasons:

- Being able to run stricter static analyzers against MoveIt2 and Nav2
- Being able to ensure build reproducibility of a particular version

I don't think we can accomplish those goals with rosdep/apt

**Reply by asimonov** [reply]
yes, i agree. mixing apt with source builds is not a great idea. apt complicates everything as long as we start installing any ros-* as it brings spurious dependencies and then everything is a mess

**Reply by franklinselva** [reply]
In that case, It would be nice in removing the ros key after handling the dependencies. This will ensure we are not meant to use apt packages instead prefer source build or report to the project for additional support in packages.

**Reply by EzraBrooks** [reply]
That's actually a great idea! I hadn't thought of that.

A very very common issue with Apt-based systems (i.e. rosdep) is that you cannot sufficiently control transitive dependency resolution. Using my primary project, MoveIt Pro, as an example: We have literally always had a source build of `moveit2` because we work on it all the time - but rosdep/Apt occasionally pull in `moveit_core` transitively and break everything. You end up needing to very carefully control your Apt dependencies if you use them at all.

**Reply by 130s** [reply]
My understanding about space-ros is limited yet. Please forgive me about that. That said, I felt installing dependency from `apt` via `rosdep` (and Catkin/Colcon framework that `rosdep` is built on top of) is underrated here, a little unnecesarrily (while I agree some of the usecases cannot be easily achieved with `rosdep`).  So I'm sharing my understanding/suggestions. Hope this helps giving some new perspectives if any.

One-liner: If there's something that cannot be done without building upstream from source, then we should source-build. If not, considering to avoid source-build more carefully is a worth-while effort.

>> It would be nice in removing the ros key after handling the dependencies. This will ensure we are not meant to use apt packages instead prefer source build or report to the project for additional support in packages.

Taking Catkin/Colcon as a framework of describing software manifest, I like `package.xml` to be a single location where all the dependencies are concisely listed in. Developer should only need to look at that file in order to know the direct dependency.

Also resources on ros.org are designed with assumption that pkgs are made in accordance with the Catkin/Colcon "framework" e.g. https://index.ros.org/packages/ that shows dependency listed within the framework.

I'd suggest 2 options that still let us keep the dependency listed in `package.xml` but prevent `rosdep` from installing certain pkgs:

- Opt-1. Keep the dependency in the manifest, but "control the installer". `rosdep` has an option to ignore certain keys ([ref (robotics.stackexchange.com#61336)](https://robotics.stackexchange.com/a/61336/60) So if the project wants to depend on e.g. `moveit_ros`:
   ```
   rosdep install --from-paths src --ignore-src --skip-keys="moveit_ros"
   ```
- Opt-2. Keep the list of dependencies in question in the manifest, but comment them out.

>>>> - Being able to run stricter static analyzers against MoveIt2 and Nav2

Isn't that something that we should encourage the upstream project to do, if it's not yet done over there?

If space_ros project wants to maintain their custom code changes, then yes I agree we (space_ros) have to do that. But if that's the case, then discussion about "`apt` or not" would've not even started. So I assume that's not the case yet?

>>>> - Being able to ensure build reproducibility of a particular version

This sounds like a valid reason to downvote using prebuilt packages IF space_ros needs to depend on specific version of upstream (e.g. nav2, moveit2) that may not be available from `apt` repo, including custom code change that space_ros makes. I don't know if there are ways to force `rosdep` to install non-latest, specific version of the pkgs listed in package.xml.

>>> yes, i agree. mixing apt with source builds is not a great idea. apt complicates everything as long as we start installing any ros-* as it brings spurious dependencies and then everything is a mess

Did you mean that developing a software using `apt` packages as a dependency in general is not a good idea? I'm curious what made you say "everything is a mess".

**Reply by franklinselva** [reply]
@130s I can see your point on this. We are already using 'rosdep'  to a certain level in handling the keys. And for option 2, I feel like it would require the majority of the ROS community to be forced to a certain way of development encouraging new standards. But my question here would be do we need to do that now or build a baseline before proposing this as a new standard?

Besides, `apt` allows pinning versions directly but `rosdep` doesn't allow that if I remember correctly. The problem with apt versions for ROS packages is they ship with two different flavours (`6.8.0-1~jammy` and `1.0.9-1jammy.20240729.010713` for instance) which seems to be due to the different CI/CD practices. [This](http://packages.ros.org/ros2/ubuntu/dists/jammy/main/binary-amd64/Packages) seems to be the only package data file being maintained at the moment when it comes to a ROS distro and it always prefers the latest version of the packages. This might be the reason why rosdep cannot pull a specific version of the ros package. And track

With these conditions in place, do we still need to lean towards apt? I am not sure but I would be happy to since we are building the source code the same way as everyone else but we will need to add a few additional stages on top of this for the safety-critic applications. 

Since SpaceROS is a meta package itself, handling source code builds at this point makes it easier to pin the dependency versions (using commit SHAs). But I am still trying to wrap my head around how we are going to propose the upstream maintainers for the strict analysers we run on top of these packages.

It ultimately boils down to the preference.

**Reply by EzraBrooks** [reply]
@130s `--skip-keys` and `--ignore-src`, while useful, do not fix the "transitive dependency" problem, unfortunately. You can only skip a root-level key, because Apt will install the transitive dependencies of your root keys no matter what. At least that's my understanding - I would be extremely happy to be wrong.

**Reply by 130s** [reply]
> You can only skip a root-level key, because Apt will install the transitive dependencies of your root keys no matter what.

Hm. I don't know if that's by design or a defect. I did a bit of test but with my understanding I also see transitive deps are getting installed. Filed a clarification https://github.com/ros-infrastructure/rosdep/issues/983

**Reply by 130s** [reply]
@franklinselva

> for option 2, I feel like it would require the majority of the ROS community to be forced to a certain way of development encouraging new standards. But my question here would be do we need to do that now or build a baseline before proposing this as a new standard?

Sorry, I'm not sure what you're exactly asking, because I don't think OPT-2 above would require effort beyond your team. To illustrate what I meant by OPT-2 is something simple like the following in package.xml e.g.:

```
:
<depend>dep_a</depend>
<!-- <depend>moveit_ros</depend> See https://github.com/space-ros/space-ros/discussions/189#discussioncomment-10398791 -->
<depend>dep_b</depend>
:
```

> Besides, `apt` allows pinning versions directly but `rosdep` doesn't allow that if I remember correctly. The problem with apt versions for ROS packages is they ship with two different flavours ... This might be the reason why rosdep cannot pull a specific version of the ros package. And track

I agree with your understanding.

I guess I forgot to add a few more options:

- OPT-C. Point to a special `apt` repo that only hosts specific version. ROS core team calls it `snapshot` ([discourse.ros.org#7705](https://discourse.ros.org/t/announcing-ros-snapshot-repositories/7705)).
   - E.g. In the snapshot made on 2024-07-03, [ros-humble-moveit-ros](http://snapshots.ros.org/humble/2024-07-03/ubuntu/pool/main/r/ros-humble-moveit-ros/) comes with `ros-humble-moveit-ros_2.5.5-1jammy.20240611.152419_amd64.deb`.
   - By swapping the config file for `apt` (on Ubuntu it's something like `/etc/apt/sources.list.d/ros2.list`), `apt` will look at this repo.
      - It's (understandably?) not so well known feature that OSRF provides. I might contribute a tutorial. It's such a useful feature.
   - ROS team may or may not recommend to rely on snapshot for production usecase. I forgot. But in my previous startup I used snapshot repo for a few years (before switching to a commercial service that provides more controlled versioning) without critical issues.
- OPT-D. Use a versioned Docker image as a basis for the development of this project. I thought someone either officially affiliated with OSRF or not was maintaining such images meticulously.
- OPT-E. Use commercial service as I mentioned in OPT-c.
   - I'm not sure if that can be an option for this project.
   - That said, Canonical was one of them that provided such a service. IDK if they are still doing that.

> With these conditions in place, do we still need to lean towards apt?

I guess the question is not about "`apt` or not" but more like "pre-built software or not", isn't it?

Sorry, I'm still slowly catching up with this seemingly huge project. I'm just providing perspectives that I don't see brought up in the record.

**Reply by EzraBrooks** [reply]
Using the ROS upstream snapshots is definitely a good idea, I agree! Even if we err on the side of trying to use Apt as little as possible, it would be a good safety net to ensure we get a particular locked date stamp of ROS. This is how I generally manage this concern on my projects at PickNik.

**Reply by Bckempa** [reply]
> > * Being able to run stricter static analyzers against MoveIt2 and Nav2
> 
> Isn't that something that we should encourage the upstream project to do, if it's not yet done over there?
> 
> If space_ros project wants to maintain their custom code changes, then yes I agree we (space_ros) have to do that. But if that's the case, then discussion about "`apt` or not" would've not even started. So I assume that's not the case yet?

Spaceflight projects are expected to have more stringent code standards than general use software; I do not mean to imply that they are better, merely more constraining. The goal of this component of the project is to provide surveillance and reporting of the state of third-party code that projects can use to decide how to proceed.

We do not expect upstream projects to adopt whichever (potentially conflicting!) standards every Space ROS user requires, rather we aim to detect when these standards are violated so a patch can be contributed upstream or provide an automated system for projects to verify their custom patch-sets bring the upstream software into compliance.

Maintaining a "space fork" is the last resort option here and would require a compelling reason why the maintenance burden should not be on the project with the violated requirement.

**Reply by Bckempa** [reply]
> I guess the question is not about "`apt` or not" but more like "pre-built software or not", isn't it?

I agree this is the real point of distinction and unless we remove the Ubuntu version assumption and go for a full Gentoo/NixOS-esque build system and construct the full OS from source there will always be another layer of pre-built/uncontrolled packages to worry about. 

While I am personally interested in that approach (especially as we look to support [more](https://github.com/space-ros/space-ros/issues/171) [architectures](https://github.com/space-ros/space-ros/issues/157#issuecomment-2291242243)), I think we can agree that it would require enough time to get there that a resolution that can pin the versions of apt packages would be a preferred near-term resolution.

Note that issue this has been discussed before: https://github.com/space-ros/docker/issues/88#issuecomment-1892930582

**Reply by franklinselva** [reply]
If we are opting for debian based OS and prebuilt-softwares, snapshots seems to be a better approach. I just had looked into the debian packages and it remains the same as other binary installation process. @130s The standards I was referring to is better explained by @Bckempa

**Reply by franklinselva** [reply]
This is a long list of ideas and perspectives and I am trying to summarize potential directions [here](https://github.com/space-ros/space-ros/discussions/189#discussioncomment-10414421)

**Reply by Bckempa** [reply]
Well [snapshots](http://wiki.ros.org/SnapshotRepository) definitely sound like what we want, but they are snapshots of the ROS packages, not the base OS packages they rely on. So this is a snapshot of the packages we want to build from source anyway and doesn't resolve out apt issue.

Even if it did, the OSRF only [promises availability for 6 months](https://discourse.ros.org/t/announcing-ros-snapshot-repositories/7705) though maybe this doesn't matter with image archiving?

**Comment by franklinselva** [comment]
Thanks for shedding lights on this. I am pretty late to understand the design decisions made and am trying to catch up to the reasons.

> * Being able to run stricter static analyzers against MoveIt2 and Nav2

Since we are running static analyzers on Nav2, Moveit2 and their third party dependencies, what are the potential next steps if the source code fails with unsafe code? In this case, are we planning to report the results to the respective repositories? or planning a fork of the work to fix it?

> * Being able to ensure build reproducibility of a particular version

I remember `apt` provides some tools to pin and hold specific version of the packages avoiding users to upgrade / downgrade to different versions. This allows specific version ids to be used and if we need SHAs of it, we could refer the data file on [packages.ros.org](http://packages.ros.org/ros2/ubuntu/dists/jammy/main/binary-amd64/Packages)

**Comment by Bckempa** [comment]
Just to echo @EzraBrooks, we're trying to move toward a reproducible source build and get as few things from the package repositories as we can get away with. There are several methods we could take to improve this, but we're working though them incrementally - starting with the build system rework issue series we are currently undertaking.

> What are the potential next steps if the source code fails with unsafe code? In this case, are we planning to report the results to the respective repositories? or planning a fork of the work to fix it?

The first and preferred option is always to work with the upstream maintainers to patch the project. If that cannot happen for whatever reason (change too specific, upstream doesn't have the bandwidth to respond, etc.) then we would consider a fork but that's a drastic step.


> I remember `apt` provides some tools to pin and hold specific version of the packages avoiding users to upgrade / downgrade to different versions. This allows specific version ids to be used and if we need SHAs of it, we could refer the data file on [packages.ros.org](http://packages.ros.org/ros2/ubuntu/dists/jammy/main/binary-amd64/Packages)

This has been on our roadmap for after the build refactor we are currently undergoing. @ xfiderek took a look at this awhile ago but it will get more attention after the current set of changes settles out. Any input or experimental results you have surrounding this are welcome.

**Reply by franklinselva** [reply]
Since I am unaware of the current set of changes in detail. It would be hard for me position my contribution. Is there any pointer for me to get to know better?

**Reply by Bckempa** [reply]
I was referring to the https://github.com/space-ros/space-ros/discussions/140 discussion and https://github.com/space-ros/space-ros/issues/107 tracking issue that I've linked a few times, but more generally you can follow the build scope by looking at the [milestone project view](https://github.com/orgs/space-ros/projects/3/views/11). The one for the next release has just been opened so it still needs some filling out.

**Comment by franklinselva** [comment]
Thanks for the explanations @Bckempa @EzraBrooks 

I think we can move this issue to discussions.

**Reply by Bckempa** [reply]
Done

**Comment by franklinselva** [comment]
A summary of the discussion so far,

1. The base requirement for Space ROS packaging is the the dependency trees of space ROS to meet certain stringent code standards which differ from public software uses.
2. To test if the dependencies are meeting the standards (currently only focusing on ROS packages and not all third party libraries), we test the source code with `IKOS` analyser to identify potential bottlenecks. Having said that, source code build contributes to this by simplifying the build testing architecture to a greater level
3. An alternative approach here is to use `rosdep` and `apt` combination to install the requirements directly. The problem with this approach is the combination always prefer the latest version of the dependencies in the tree.
4. A modified approach to the previous case is using ROS  `snapshots` which actually lifts the problem of packages being synced frequently. We can decide the sync interval of the packages to a much better level.
5. The approaches mentioned in `3.` and `4.` are prebuilt-packages and do not ship a reference of source code which complicates the build-test process. To overcome this issue, the process of getting to the date of the snapshot should be from the source code and not the other way around. This will help us meet the build test process easily with a slightly modified build architecture.
6. Getting back to the transitive dependencies of the dependency tree, I can see both approaches in `3.` and `4.` behave similarly. We will need to keep a list of excluded dependencies to prune the dependency tree so we can use it with rosdep's `--skip-key` or apt's exclude dependency mechanism.
7. Finally, we are planning to scale the packaging to different platforms and architectures which leaves this point as one of the decision criteria in the approaches moving forward
8. Getting back to the build test, if a dependency fails to meet the standard, we are expected to report it to the upstream maintainer for a fix / send out a patch to the upstream repository. This will help us avoid a fork and maintain it to an extend depending on the scale of the issue. If the issue stands at a large scale, we may need to fork the work, patch and maintain them periodically. This redundant the prebuilt-packages.

If there is any point missing in the above list, we could add it.

**Comment by EzraBrooks** [comment]
Just to make a point of clarification on one of my remarks about transitive dependencies and disentangle that from the rest of the context of that thread:

It’s notoriously difficult to exclude transitive dependencies of Apt packages (often impossible and you’ll just get a “held broken packages” error that borks your Apt). It’s also difficult to pin the versions of a large set of dependencies - @eholum and I wrote a pretty nasty script when we were working on OSAM-1 to manage the recursive mess that is “replacing a set of fixed package versions with a new set of fixed package versions”. Due to rosdep being a fairly thin wrapper around Apt on Debian-based distros, it inherits these issues.

A practical example I experienced in the past year was depending upon a MoveIt plugin whose Apt package depended upon part of MoveIt. As a result, even though I had a source build of MoveIt, rosdep asked Apt to install the plugin and Apt happily installed a conflicting binary package of MoveIt due to the transitive dependency.

These Apt shortcomings are arguably by design - the Debian project is strictly adherent to the principle that “every package in the distribution will always work with every other package” - but I think we all know that never really works in practice :wink:

**Comment by asimonov** [comment]
I support what @EzraBrooks and @Bckempa said about philosophy of building [ROS] dependencies from source as much as possible, keeping track of exact versions we use.

Learning and fighting apt/rosdep is a distraction, in my opinion. It hides exact versions, it does not provide source code, it adds extra misdirection and complexity making debug much harder.
We should only use apt for non-ROS dependencies, something like opencv for example.

Having well tracked source build will allow us to run code analysis and port to non ubuntu/apt architectures. Because a lot of dependencies are localised in the source build.

Another consideration - it is much easier to go the other way - starting with well designed source build and convert parts of it to use apt, if a user wants. It is much harder to go the other way. We should solve the harder problem.

**Reply by franklinselva** [reply]
+1
